### PR TITLE
Task/jenkins 43741 Improved test coverage of blueocean-github-pipeline

### PR DESCRIPTION
**Pull request by vivek**
# Description

See [JENKINS-43741](https://issues.jenkins-ci.org/browse/JENKINS-43741).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

**Comment by vivek** [comment]
@i386 @michaelneale I ended up using Wiremock with powermock/mochito, was much easier to mock GitHub than doing it powermock/Mockito way.

Unfortunately, I realized after doing most of GitHub mocking that Wiremock needs guava 18. We can't upgrade to it as jenkins core enforces guava 11.0.1 and then guava has incompatible changes across versions. For example `Iterators.skip()` was removed starting version 13! This has been refactored in this PR though. For now all tests pass and things look ok but this is something to watch out for in order to use wire mock.

There is incompatibility between powermock and Jacoco - Jacoco ignores tests that use power mock, but this was mitigated by changing Jacoco configuration  to do offline analysis (using jacoco javaagent), so the instrumentation happens early one and report is generated during pre-package phase. IDEA also does better job reporting coverage, useful as data point.

All in all. Wire mock, makes GitHub mocking manageable, however due to guava incompatible versions we have to watch out for incompatibility. Side effect of it is only applicable during test runs and probably `mvn hpi:run`, I think this includes test dependencies as well.

If we decide to not use wire mock, then the only alternative is to use GitHub directly, thats not bad either but that has it's own complexity.

**Comment by i386** [comment]
> With these changes creation has 80% coverage

@vivek you are a champion!

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java [hunk 1]
```diff
@@ -340,16 +339,6 @@ static void validateCredentialId(String credentialId,  String apiUrl) throws IOE
         }
     }
 
-    private static void deleteOnError(AbstractFolder item){
```
Comment (i386) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java]: Why was this deleted?

Comment (vivek) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java]: Yes, not used.

**Comment by vivek** [comment]
It got decent coverage now > 70%, at least for most happy paths. According to IDEA it's 78%, Jacoco shows 71%. The difference could be because of not too many lines or some config related thing. Both show 100% class coverage.

IDEA:

image

Jacoco:

image

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java (hunk 1 above)
Comment (michaelneale) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java]: @i386 wasn't this what was put in to clean up a partially created org folder?

**Comment by michaelneale** [comment]
@vivek yes hpi:run does include test scope - so that could be a problem. Guava dependencies seem dangerous territories. So what is solution? 

This is looking great so far, good coverage!

File: blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java (hunk 1 above)
Comment (i386) [inline blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java]: @michaelneale it does so in `cleanupOnError`

**Comment by vivek** [comment]
@michaelneale This is fixed. mvn hpi:run is fixed to resolve dependency with runtime scope. It's better anyways as it runs `hpi:run` closer to actual jenkins deployment. If anyone wants to run with `test` dependency resolution strategy then they can  use `mvn hpi:run -Dhpi.dependencyResolution=test`.

**Comment by vivek** [comment]
ATH passes https://ci.blueocean.io/blue/organizations/jenkins/blueocean/detail/task%2FJENKINS-43741/8/pipeline

**Comment by michaelneale** [comment]
Seems good to me :bee:

**Comment by tfennelly** [comment]
@vivek hey Vivek, these changes have resulted in `mvn -T 1C clean test-compile install -DskipTests` not working for me. If I got to the commit before this merge it works fine.

The error I get is the following ...

```
[INFO] Blue Ocean Parent .................................. SUCCESS [ 10.502 s]
[INFO] Common API for Blue Ocean .......................... FAILURE [ 13.111 s]
[INFO] REST API for Blue Ocean ............................ SKIPPED
[INFO] i18n for Blue Ocean ................................ SKIPPED
[INFO] Web for Blue Ocean ................................. SKIPPED
[INFO] JWT for Blue Ocean ................................. SKIPPED
[INFO] REST Implementation for Blue Ocean ................. SKIPPED
[INFO] Pipeline REST API for Blue Ocean ................... SKIPPED
[INFO] Events API for Blue Ocean .......................... SKIPPED
[INFO] Dashboard for Blue Ocean ........................... SKIPPED
[INFO] Personalization for Blue Ocean ..................... SKIPPED
[INFO] Config API for Blue Ocean .......................... SKIPPED
[INFO] GitHub Pipeline for Blue Ocean ..................... SKIPPED
[INFO] Git Pipeline for Blue Ocean ........................ SKIPPED
[INFO] Blue Ocean ......................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 27.986 s (Wall Clock)
[INFO] Finished at: 2017-05-10T12:45:43+01:00
[INFO] Final Memory: 149M/1035M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.7.9:instrument (default-instrument) on project blueocean-commons: Unable to instrument file. Error while instrumenting class /Users/tfennelly/projects/blueocean/blueocean-commons/target/classes/io/jenkins/blueocean/commons/BlueOceanConfigProperties.class. Class io/jenkins/blueocean/commons/BlueOceanConfigProperties is already instrumented. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :blueocean-commons
```

**Comment by vivek** [comment]
@tfennelly Jacoco is doing instrumentation during `process-classes` phase. Failure is because `process-classes` is invoked first during `test-compile` and then during `install`, Jacoco is trying instrumentation second time for the same class and fails. I tried few things but it didn't work. As a work around you can always disable Jacoco, run your mvn with `-Djacoco.skip=true` option.
